autoflip_calculators/active_speaker_detection/lip_track_calculator.cc

```diff
@@ -0,0 +1,261 @@
+  options_ = cc->Options<LipTrackCalculatorOptions>();
+
+  return ::mediapipe::OkStatus();
+}
+
+::mediapipe::Status LipTrackCalculator::Process(
+    mediapipe::CalculatorContext* cc) {
+  auto output_bbox = ::absl::make_unique<std::vector<NormalizedRect>>();
+  std::vector<NormalizedRect> cur_face_bbox;
+  std::map<int32, std::deque<float>> cur_face_statistics;
+  std::vector<float> statistics;
+  
+  cv::Mat frame;
+  frame = mediapipe::formats::MatView(
+      &cc->Inputs().Tag(kInputVideo).Get<ImageFrame>());
+  frame_width_ = frame.cols;
+  frame_height_ = frame.rows;
+
+  const auto& input_landmark_lists = 
+        cc->Inputs().Tag(kInputLandmark).Get<std::vector<NormalizedLandmarkList>>();
+  const auto& input_face_bbox =
+        cc->Inputs().Tag(kInputROI).Get<std::vector<NormalizedRect>>();
+  
+  GetStatistics(input_landmark_lists, &statistics);
+
+  for (int i = 0; i < input_face_bbox.size(); ++i) {
+    auto& bbox = input_face_bbox[i];
+    cur_face_bbox.push_back(bbox);
+    int32 cur_face_idx = cur_face_bbox.size() - 1;
+
+    int previous_face_idx = MatchFace(bbox);
+    cur_face_statistics.insert(std::pair< int32, std::deque<float> >(cur_face_idx, std::deque<float>()));
+    if (previous_face_idx != -1) {
+      for (auto& value : face_statistics_[previous_face_idx])
+        cur_face_statistics[cur_face_idx].push_back(value);
+      while (cur_face_statistics[cur_face_idx].size() > options_.frame_history() - 1)
+        cur_face_statistics[cur_face_idx].pop_front();
+    }
+    else {
+      cur_face_statistics[cur_face_idx].push_back(statistics[i]);
+    }
+    if (IsActiveSpeaker(cur_face_statistics[cur_face_idx]))
+      output_bbox->push_back(bbox);
+  }
+
+  // Update the history
+  face_bbox_ = cur_face_bbox;
+  face_statistics_ = cur_face_statistics;
+
+  cc->Outputs().Tag(kOutputROI).Add(output_bbox.release(), cc->InputTimestamp());
+
+  return ::mediapipe::OkStatus();
+} 
+
+float LipTrackCalculator:: GetDistance(const NormalizedLandmark& mark_1,
+                                      const NormalizedLandmark& mark_2) {
+  return std::sqrt(std::pow((mark_1.x()-mark_2.x())*frame_width_, 2) 
+  + std::pow((mark_1.y()-mark_2.y())*frame_height_, 2));
+}
+
+void LipTrackCalculator:: GetStatistics(const std::vector<NormalizedLandmarkList>& landmark_lists, 
+                    std::vector<float>* lip_statistics) {
+  float mouth_width;
+  float mouth_height = 0.0f;
+  for (const auto& landmark_list : landmark_lists) {
+    if (landmark_list.landmark_size() < kFaceMeshLandmarks)
+      continue;
+    mouth_width = GetDistance(landmark_list.landmark(kLipLeftInnerCornerIdx),
+                              landmark_list.landmark(kLipRightInnerCornerIdx));
+    for (auto i = 0; i < kLipUpperIdx.size(); ++i) {
+      mouth_height += GetDistance(landmark_list.landmark(kLipUpperIdx[i]),
+                              landmark_list.landmark(kLipLowerIdx[i]));
+    }
+    mouth_height /= (float)kLipUpperIdx.size();
```

Comment:
Can you add a small comment here on why this divide is needed?

Reply:
We can remove it. It just averages the mouth_height, since we calculated 3 distances.